server/src/main/java/org/opensearch/search/aggregations/AggregatorBase.java

```diff
@@ -216,6 +220,10 @@ protected void preGetSubLeafCollectors(LeafReaderContext ctx) throws IOException
      */
     protected void doPreCollection() throws IOException {}
 
+    protected boolean tryPrecomputeAggregationForLeaf(LeafReaderContext ctx) throws IOException {
```

Comment:
Propose adding some comment: This should be a segment-level self-contained way of computing the same aggregation results before collecting and return early. Please be careful about updating any Aggregator class states from it, as that might be applicable to subsequent segments.